File: tp3/Ej6/ej6.m
cutoff = 10;
order = 2;

im=double(imread('tun.jpg'));

[x, y]=size(im);


%Butterworth filter
A=zeros(x,y);
for i=1:x
    for j=1:y
        A(i,j)=(((i-x/2).^2+(j-y/2).^2)).^(.5);
        H(i,j)=1/(1+((d/A(i,j))^(2*n)));
    end
end

alphaL=.0999;
aplhaH=1.01;

H=((aplhaH-alphaL).*H)+alphaL;
H=1-H;

%log
im_l=log2(1+im);

%DFT
im_f=fft2(im_l);

%Filter
im_nf=H.*im_f;

%Inverse DFT
im_n=abs(ifft2(im_nf));

%Inverse log
im_e=exp(im_n);

figure, imshow((im_e),[])

File: tp3/Ej6/ej6.py
import numpy as np
import cv2

cutoff = 10
order = 2

image = cv2.imread('../image_sources/tun.jpg')

gray_image = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY).astype(np.float64)

x, y = gray_image.shape

# Add a small offset to avoid division by zero
epsilon = 1e-8  # Small positive constant
gray_image_offset = gray_image + epsilon

# Calculate the logarithm
image_logued = np.log(1 + gray_image_offset)

# DFT
image_dft = np.fft.fft2(image_logued)

# Filter
A = np.zeros((x, y))
H = np.zeros((x, y))
d = cutoff
n = order

for i in range(x):
    for j in range(y):
        A[i, j] = np.sqrt((i - x / 2) ** 2 + (j - y / 2) ** 2)
        if A[i, j] == 0:
            H[i, j] = 0  # Handle division by zero by setting H to a default value (0 in this case)
        else:
            H[i, j] = 1 / (1 + ((d / A[i, j]) ** (2 * n)))

Yh = 0.0999
Yl = 1.01

H = ((Yl - Yh) * H) + Yh
H = 1 - H

image_f = H * image_dft

# Inverse DFT
image_n = np.abs(np.fft.ifft2(image_f))

# Inverse log
image_final = np.exp(image_n)

# Apply the full dynamic range display (similar to MATLAB's imshow(_, []))
min_val = np.min(image_final)
max_val = np.max(image_final)
image_adjusted = ((image_final - min_val) / (max_val - min_val) * 255).astype(np.uint8)

# Display the final image
cv2.imwrite('./results/homomorphicImageLn.jpg', image_adjusted)
